src/auth_context.rs

```diff
@@ -0,0 +1,172 @@
+//! API for authenticating peer
+//! Based on https://grpc.github.io/grpc/core/md_doc_server_side_auth.html
+
+use std::ffi::CStr;
+use std::marker::PhantomData;
+
+use crate::grpc_sys::{
+    self, GrpcAuthContext, GrpcAuthProperty, GrpcAuthPropertyIterator, GrpcCall,
+};
+
+/// To perform server-side authentication, gRPC exposes the authentication context
+/// for each call. The context exposes important authentication-related information
+/// about the RPC such as the type of security/authentication type being used and
+/// the peer identity.
+///
+/// The authentication context is structured as a multi-map of key-value pairs -
+/// the auth properties. In addition to that, for authenticated RPCs, the set of
+/// properties corresponding to a selected key will represent the verified identity
+/// of the caller - the peer identity.
+///
+/// The contents of the auth properties are populated by an auth interceptor within
+/// gRPC Core. The interceptor also chooses which property key will act as the peer
+/// identity (e.g. for client certificate authentication this property will be
+/// `x509_common_name` or `x509_subject_alternative_name`).
+pub struct AuthContext<'a> {
+    ctx: *mut GrpcAuthContext,
+    _lifetime: PhantomData<&'a GrpcAuthContext>,
+}
+
+/// Binding to gRPC Core AuthContext
+/// If the server binds in non-secure mode, all functions will have a behaviour
+/// consistent with client not being authenticated, for ease of use and speed
+impl<'a> AuthContext<'a> {
+    /// Will be created even if grpc_call_auth_context is null
+    pub(crate) unsafe fn from_call_ptr(call: *mut GrpcCall) -> Self {
```

Comment:
As documented above: *If the server binds in non-secure mode, all functions will have a behaviour consistent with client not being authenticated, for ease of use and speed.*

I've been using this API for almost a year now, and it's indeed much more practical than having an option there. It is anyway very unlikely that you're going to use this function if the server was not configured to run in secure mode.

Reply:
I don't think it makes any significant performance difference here.

If the connection is insecure, `AuthContext` should not be used in the first place. There is no point to have a consistent behavior when they should not.

Reply:
I agree it does not make a significant performance difference, I'm essentially advocating for the ease of use of this interface.

As you pointed out, if the connection is insecure, `auth_context` should not be used in the first place, so I feel that having to deal with an Option when calling it would damage the interface.

And in the scenario where your code can bind in both secure and non secure mode, you probably want it to behave as non-authenticated when going through an insecure path.

In the unlikely event that one would want any other behavior, they can call `is_auth_context_present` to differentiate.

Reply:
> ...deal with an Option when calling it would damage the interface.

`Option` makes it very clear that `auth_context` may not exist. It's the *correct* and rusty way to represent such semantic in rust.